### PR TITLE
CLOUDSTACK-8721: Fixed Setting details of VM through API results in removal of all other details except the one passed in API

### PR DESCRIPTION
**Pull request by anshul1886**

### CODE REVIEW

**Comment by anshul1886** [comment]
cloudstack-pull-requests #945 failure is because of some random error in git fetching

**Comment by remibergsma** [comment]
@anshul1886 I want to test this but haven't used the 'details' feature before. In what format am I supposed to send the key/value pairs? If you have a CloudMonkey or Marvin example, that would be great. I'll then verify it in my lab. Thanks!

**Comment by anshul1886** [comment]
@remibergsma Using cloudmonkey we can pass as shown below 

update virtualmachine id=00f19661-da81-40b2-a07c-e46967421195 details[0].keyboard=fr

Here detail which we want to update has key "keyboard" and its value is "fr" 

And using API

http://localhost:8096/api/?command=updateVirtualMachine&id=00f19661-da81-40b2-a07c-e46967421195&details[0].keyboard="fr"

**Comment by remibergsma** [comment]
Thanks @anshul1886 will test it soon.

**Comment by remibergsma** [comment]
@anshul1886 Tested your code, works just fine! LGTM.

The only thing I noticed: excisting key/values are removed, then added again. So the net result is OK. It just inserts them again (the auto-increment goes up) but that wasn't introduced by your code as far as I can see.

File: api/src/org/apache/cloudstack/api/command/user/vm/UpdateVMCmd.java [hunk 1]
```diff
@@ -136,13 +136,13 @@ public String getInstanceName() {
         return instanceName;
     }
 
-    public Map getDetails() {
+    public Map<String, String> getDetails() {
```
Comment (DaanHoogland) [inline api/src/org/apache/cloudstack/api/command/user/vm/UpdateVMCmd.java]: how about taking the opportunity to change this to 
    public Map<String, String> getDetails() {
        if (this.details == null || this.details.isEmpty()) {
            return null;
        }
        return details;
    }
Is there any reason to pass through the crazy construct of getting the pointer to the first element and casting it back to a map? It seems like very unsafe. It becomes even more unsafe when using an instance of a generic type.

Comment (DaanHoogland) [inline api/src/org/apache/cloudstack/api/command/user/vm/UpdateVMCmd.java]: @bhaisaab can you comment on this?

Comment (DaanHoogland) [inline api/src/org/apache/cloudstack/api/command/user/vm/UpdateVMCmd.java]: come to think of it
```
    public Map<String, String> getDetails() {
        return details;
    }
```
should even do

Comment (bhaisaab) [inline api/src/org/apache/cloudstack/api/command/user/vm/UpdateVMCmd.java]: The CommandType.MAP API args are usually map of object returning objects; in practice AFAIR all map args are string->string. Either way, I don't understand this change in getDetails(), this at least will result in returning a new Map, so likely the author wants their details to be sort of read-only? If so, why not use final?

File: server/src/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -2188,7 +2188,13 @@ public UserVm updateVirtualMachine(UpdateVMCmd cmd) throws ResourceUnavailableEx
         }
 
         if (details != null && !details.isEmpty()) {
-            vmInstance.setDetails(details);
+            _vmDao.loadDetails(vmInstance);
```
Comment (bhaisaab) [inline server/src/com/cloud/vm/UserVmManagerImpl.java]: I think all we need is here, the side effect is that now users won't be able to remove any details; only add/update details.

**Comment by bhaisaab** [comment]
Ping, update on this?

**Comment by anshul1886** [comment]
@bhaisaab What else is needed here?

File: api/src/org/apache/cloudstack/api/command/user/vm/UpdateVMCmd.java (hunk 1 above)
Comment (anshul1886) [inline api/src/org/apache/cloudstack/api/command/user/vm/UpdateVMCmd.java]: This is to avoid long nested type casting before using details.

**Comment by bhaisaab** [comment]
@anshul1886 can you read and reply to some of the comments?

File: server/src/com/cloud/vm/UserVmManagerImpl.java (hunk 2 above)
Comment (anshul1886) [inline server/src/com/cloud/vm/UserVmManagerImpl.java]: When do we need to remove detail? Even if needed then those values can be updated to get the desired effect. Many details are for internal purposes only and user is not even aware of them.

**Comment by anshul1886** [comment]
@bhaisaab Lost in discussion and got only end result that all we need is here.

**Comment by bhaisaab** [comment]
@anshul1886 the only concern here is that this sort of breaks the API behaviour as in the past the details would set details (removing old details).

LGTM; we can merge once anyone else can review/test this.

**Comment by anshul1886** [comment]
@remibergsma @bhaisaab Remi has already tested this and has given the LGTM.

**Comment by bhaisaab** [comment]
ah, I see merging now :)

**Pull request merged**